spring-integration-core/src/main/java/org/springframework/integration/graph/GraphBuilder.java

```diff
@@ -0,0 +1,138 @@
+		Map<String, EventDrivenConsumer> eventDriven = this.applicationContext
+				.getBeansOfType(EventDrivenConsumer.class);
+		for (AbstractMessageChannel channel : channels.values()) {
+			map.put(channel.getComponentName(), new MessageChannelInfo(channel));
+		}
+		Collection<LinkNode> links = new ArrayList<LinkNode>();
+		for (SourcePollingChannelAdapter adapter : spcas.values()) {
+			MessageSourceInfo sourceInfo = sourceInfo(adapter);
+			map.put(adapter.getComponentName(), sourceInfo);
+			links.add(new LinkNode(sourceInfo.getNodeId(), map.get(sourceInfo.getOutput()).getNodeId()));
+		}
+		for (MessagingGatewaySupport gateway : gateways.values()) {
+			MessageGatewayInfo gatewayInfo = gatewayInfo(gateway);
+			map.put(gateway.getComponentName(), gatewayInfo);
+			links.add(new LinkNode(gatewayInfo.getNodeId(), map.get(gatewayInfo.getOutput()).getNodeId()));
+		}
+		for (MessageProducerSupport producer : producers.values()) {
+			MessageProducerInfo producerInfo = producerInfo(producer);
+			map.put(producer.getComponentName(), producerInfo);
+			links.add(new LinkNode(producerInfo.getNodeId(), map.get(producerInfo.getOutput()).getNodeId()));
+		}
+		for (PollingConsumer consumer : polling.values()) {
+			MessageHandlerInfo handlerInfo = handlerInfo(consumer);
+			map.put(consumer.getComponentName(), handlerInfo);
+			links.add(new LinkNode(map.get(handlerInfo.getInput()).getNodeId(), handlerInfo.getNodeId()));
+			if (handlerInfo.getOutput() != null) {
+				links.add(new LinkNode(handlerInfo.getNodeId(), map.get(handlerInfo.getOutput()).getNodeId()));
+			}
+		}
+		for (EventDrivenConsumer consumer : eventDriven.values()) {
+			MessageHandlerInfo handlerInfo = handlerInfo(consumer);
+			map.put(consumer.getComponentName(), handlerInfo);
+			links.add(new LinkNode(map.get(handlerInfo.getInput()).getNodeId(), handlerInfo.getNodeId()));
+			if (handlerInfo.getOutput() != null) {
+				links.add(new LinkNode(handlerInfo.getNodeId(), map.get(handlerInfo.getOutput()).getNodeId()));
+			}
+		}
+		// create the nodes
+		Collection<GraphNode> nodes = new ArrayList<GraphNode>();
+		for (Entry<String, NodeInfo> entry : map.entrySet()) {
+			NodeInfo value = entry.getValue();
+			nodes.add(new GraphNode(entry.getKey(), value.getComponentType(), value));
+		}
+		this.graph = new Graph(nodes, links);
+	}
+
+	private MessageGatewayInfo gatewayInfo(MessagingGatewaySupport gateway) {
+		return new MessageGatewayInfo(gateway, gateway.getRequestChannelName());
+	}
+
+	private MessageProducerInfo producerInfo(MessageProducerSupport producer) {
+		return new MessageProducerInfo(producer, producer.getOutputChannelName());
+	}
+
+	private MessageSourceInfo sourceInfo(SourcePollingChannelAdapter adapter) {
+		DirectFieldAccessor accessor = new DirectFieldAccessor(adapter);
+		AbstractMessageSource<?> source = (AbstractMessageSource<?>) accessor.getPropertyValue("source");
```

Comment:
Yes; originally I didn't want to expose but found with the handler I really had no choice. Will fix.